lib/beaker-hostgenerator/generator.rb

```diff
@@ -32,43 +37,38 @@ def generate(tokens, options)
           next
         end
 
-        node_info = __parse_node_info_token(token)
+        node_info = parse_node_info_token(token)
+
+        # Build node host name
+        platform = "#{ostype}-#{node_info['bits']}"
+        host_name = "#{platform}-#{nodeid[ostype]}"
 
+        node_info['platform'] = platform
         node_info['ostype'] = ostype
         node_info['nodeid'] = nodeid[ostype]
 
-        host_config = base_host_config()
-
-        [:pe_dir, :pe_ver, :pe_upgrade_dir, :pe_upgrade_ver].each do |option|
-          if options[option]
-            host_config[option.to_s] = options[option]
-          end
-        end
+        host_config = base_host_config(options)
 
+        # Delegate to the hypervisor
         hypervisor = BeakerHostGenerator::Hypervisor.create(node_info, options)
-        host_name, host_config =
-                   hypervisor.generate_node(node_info, host_config, bhg_version)
+        host_config = hypervisor.generate_node(node_info, host_config, bhg_version)
         config['CONFIG'].deep_merge!(hypervisor.global_config())
 
+        # Merge in any arbitrary key-value host settings. Treat the 'hostname'
+        # setting specially, and don't merge it in as an arbitrary setting.
+        arbitrary_settings = node_info['host_settings']
+        if arbitrary_settings['hostname']
+          host_name = arbitrary_settings['hostname']
+          arbitrary_settings.delete('hostname')
+        end
+        host_config.merge!(arbitrary_settings)
+
         if PE_USE_WIN32 && ostype =~ /windows/ && node_info['bits'] == "64"
           host_config['ruby_arch'] = 'x86'
           host_config['install_32'] = true
         end
 
-        if not options[:disable_default_role]
-          host_config['roles'] = ['agent']
-        else
-          host_config['roles'] = []
-        end
-
-        host_config['roles'].concat __generate_host_roles(node_info)
-        host_config['roles'].uniq!
-
-        if not options[:disable_role_config]
-          host_config['roles'].each do |role|
-            host_config.deep_merge! __get_role_config(role)
-          end
-        end
+        __generate_host_roles!(host_config, node_info, options)
```

Comment:
Like I said in my other comment, I tried many times to decompose this cleanly, but I didn't like any of them.
There's too many local variables weaved through this method, and I believe it could be rearranged to allow for better decomposition, but I didn't want to do too much all at once.

I did extract all the role code into its own method below, so that helps a bit.